src/ln/channelmonitor.rs

```diff
@@ -1526,6 +1848,74 @@ impl ChannelMonitor {
+		let mut htlc_updated = Vec::new();
+
+		let commitment_number = 0xffffffffffff - ((((tx.input[0].sequence as u64 & 0xffffff) << 3*8) | (tx.lock_time as u64 & 0xffffff)) ^ self.commitment_transaction_number_obscure_factor);
+		if commitment_number >= self.get_min_seen_secret() {
+			if let Some(ref current_local_signed_commitment_tx) = self.current_local_signed_commitment_tx {
+				for htlc_output in &current_local_signed_commitment_tx.htlc_outputs {
+					htlc_updated.push((htlc_output.3.clone(), None, htlc_output.0.payment_hash.clone()))
+				}
+			}
+			if let Some(ref prev_local_signed_commitment_tx) = self.prev_local_signed_commitment_tx {
+				for htlc_output in &prev_local_signed_commitment_tx.htlc_outputs {
+					htlc_updated.push((htlc_output.3.clone(), None, htlc_output.0.payment_hash.clone()))
+				}
+			}
+			// No need to check remote_claimabe_outpoints, symmetric HTLCSource must be present as per-htlc data on local commitment tx
+		} else if tx.input.len() > 0{
+			for input in &tx.input {
+				let mut payment_data: (Option<HTLCSource>, Option<PaymentPreimage>, Option<PaymentHash>) =  (None, None, None);
+				if let Some(ref current_local_signed_commitment_tx) = self.current_local_signed_commitment_tx {
+					if input.previous_output.txid == current_local_signed_commitment_tx.txid {
+						for htlc_output in &current_local_signed_commitment_tx.htlc_outputs {
+							if input.previous_output.vout == htlc_output.0.transaction_output_index {
+								payment_data = (htlc_output.3.clone(), None, Some(htlc_output.0.payment_hash.clone()));
+							}
+						}
+					}
+				}
+				if let Some(ref prev_local_signed_commitment_tx) = self.prev_local_signed_commitment_tx {
+					if input.previous_output.txid == prev_local_signed_commitment_tx.txid {
+						for htlc_output in &prev_local_signed_commitment_tx.htlc_outputs {
+							if input.previous_output.vout == htlc_output.0.transaction_output_index {
+								payment_data = (htlc_output.3.clone(), None, Some(htlc_output.0.payment_hash.clone()));
+							}
+						}
+					}
+				}
+				if let Some(htlc_outputs) = self.remote_claimable_outpoints.get(&input.previous_output.txid) {
+					for htlc_output in htlc_outputs {
+						if input.previous_output.vout == htlc_output.0.transaction_output_index {
+							payment_data = (htlc_output.1.clone(), None, Some(htlc_output.0.payment_hash.clone()));
+						}
+					}
+				}
+				// If tx isn't solving htlc output from local/remote commitment tx and htlc isn't outbound we don't need
+				// to broadcast solving backward
+				if payment_data.0.is_some() && payment_data.2.is_some() {
+					let mut payment_preimage = PaymentPreimage([0; 32]);
+					let mut preimage = None;
+					if input.witness.len() == 5 && input.witness[4].len() == ACCEPTED_HTLC_SCRIPT_WEIGHT {
+						for (arr, vec) in payment_preimage.0.iter_mut().zip(tx.input[0].witness[3].iter()) {
```

Comment:
I think you want payment_preimage.0[..].copy_from_slice().

Reply:
Ah I like to complicate my life! corrected